client/session.go

```diff
@@ -1898,21 +1924,24 @@ func (r *blocksResult) addBlockFromPeer(id ts.ID, block *rpc.Block) error {
 	switch {
 	case segments.Merged != nil:
 		// Unmerged, can insert directly into a single block
-		result.Reset(start, ts.Segment{
-			Head: segments.Merged.Head,
-			Tail: segments.Merged.Tail,
-		})
+		result.Reset(start, r.segmentForBlock(segments.Merged))
 
 	case segments.Unmerged != nil:
 		// Must merge to provide a single block
+		readerPool := r.blockOpts.SegmentReaderPool()
 		readers := make([]io.Reader, len(segments.Unmerged))
 		for i := range segments.Unmerged {
-			readers[i] = xio.NewSegmentReader(ts.Segment{
-				Head: segments.Unmerged[i].Head,
-				Tail: segments.Unmerged[i].Tail,
-			})
+			reader := readerPool.Get()
+			reader.Reset(r.segmentForBlock(segments.Unmerged[i]))
+			readers[i] = reader
 		}
+
 		encoder, err := r.mergeReaders(start, readers)
+		for _, reader := range readers {
+			// Close each reader
+			reader.(xio.SegmentReader).Close()
```

Comment:
Because `readers` is an slice of `[]io.Reader` which `mergeReaders` takes, not a slice of `[]xio.SegmentReader` and the `io.Reader` interface has no `Close()` method on it.